User_Service/Interfaces/Controllers/controller.go

```diff
@@ -0,0 +1,312 @@
+func NewController(repo *database.Repository) *Controller {
+	return &Controller{Repo: repo}
+}
+
+func (ctrl *Controller) Register(c *gin.Context) {
+
+	var reg_data entity.User
+
+	if err := c.ShouldBindJSON(&reg_data); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	err := ctrl.Repo.WhereUsername(reg_data.Username, &reg_data)
+	if err == nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Username already exist"})
+		return
+	}
+	// Check if the role exists
+	var role entity.Role
+	err1 := ctrl.Repo.WhereRoleID(reg_data.Role_id, &role)
+	if err1 != nil {
+
+		if reg_data.Role_id == "1" {
+
+			role.Role_id = reg_data.Role_id
+			role.Role_type = "Customer"
+
+			ctrl.Repo.CreateRole(&role)
+
+		} else if reg_data.Role_id == "2" {
+
+			role.Role_id = reg_data.Role_id
+			role.Role_type = "Delivery driver"
+
+			ctrl.Repo.CreateRole(&role)
+
+		} else {
+			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid role ID"})
+			return
+		}
+
+	}
+
+	err = ctrl.Repo.CreateUser(&reg_data)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err})
+		return
+	}
+
+	log.Printf("User created successfully: %+v", reg_data)
+
+	userWithRole, err := ctrl.Repo.LoadUserWithRole(reg_data.User_id)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": reg_data.User_id})
+		return
+	}
+
+	c.JSON(http.StatusCreated, userWithRole)
+}
+
+func (ctrl *Controller) Login(c *gin.Context) {
+
+	var input struct {
+		Username string `json:"username" binding:"required"`
+		Password string `json:"password" binding:"required"`
+	}
+
+	if err := c.ShouldBindJSON(&input); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	var user entity.User
+	err := ctrl.Repo.WhereUsername(input.Username, &user)
+	if err != nil {
+		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid credentials"})
+		return
+	}
+
+	if user.Password != input.Password {
+		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid password"})
+		return
+	}
+
+	token, err := utils.GenerateJWT(user.Username)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": "Could not generate token"})
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{
+		"token":      token,
+		"expires_at": time.Now().Add(24 * time.Hour).Unix(),
+	})
+
+}
+
+func (ctrl *Controller) Get_user(c *gin.Context) {
+
+	var user_data []entity.User
+	user_data, err := ctrl.Repo.Preload_in_order()
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+		return
+	}
+	c.JSON(http.StatusOK, user_data)
+
+}
+func (ctrl *Controller) Get_role(c *gin.Context) {
+	var user_data []entity.Role
+	user_data, err := ctrl.Repo.Role_in_Asc_order()
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+		return
+	}
+	c.JSON(http.StatusOK, user_data)
+}
+
+func (ctrl *Controller) Update_Role(c *gin.Context) {
```

Comment:
are we updating user role here? if yes, why do we have a separate call for just updating role?